app/src/main/java/com/github/polybooks/core/databaseImpl/OLBookDatabase.kt

```diff
@@ -0,0 +1,230 @@
+
+import android.annotation.SuppressLint
+import com.github.polybooks.core.Book
+import com.github.polybooks.core.Interest
+import com.github.polybooks.core.database.*
+import com.github.polybooks.core.database.BookOrdering.*
+import com.github.polybooks.utils.url2json
+import com.google.gson.JsonArray
+import com.google.gson.JsonElement
+import com.google.gson.JsonObject
+import java.io.FileNotFoundException
+import java.text.SimpleDateFormat
+import java.util.*
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CompletionException
+
+
+private const val TITLE_FIELD_NAME = "full_title"
+private const val AUTHORS_FIELD_NAME = "authors"
+private const val FORMAT_FIELD_NAME = "physical_format"
+private const val ISBN13_FIELD_NAME = "isbn_13"
+private const val PUBLISHER_FIELD_NAME = "publishers"
+private const val PUBLISH_DATE_FIELD_NAME = "publish_date"
+
+private const val DATE_FORMAT = "MMM dd, yyyy"
+private const val ISBN13_FORMAT = """[0-9]{13}"""
+private const val ISBN10_FORMAT = """[0-9]{9}[0-9X]"""
+private const val ISBN_FORMAT = """($ISBN10_FORMAT)|($ISBN13_FORMAT)"""
+
+private const val OL_BASE_ADDR = """https://openlibrary.org"""
+
+/**
+ * An implementation of a book database based on the Open Library online database
+ * */
+@SuppressLint("NewApi")
+class OLBookDatabase : BookDatabase {
+
+    override fun queryBooks(): BookQuery = OLBookQuery()
+
+    private inner class OLBookQuery : BookQuery {
+
+        private var ordering = DEFAULT
+
+        private var empty : Boolean = false
+        private var title : String? = null
+        private var isbn : String? = null
+
+        override fun onlyIncludeInterests(interests: Collection<Interest>): BookQuery {
+            System.err.println("Warning: onlyIncludeInterest not fully implemented for OLBookQuery")
+            this.empty = true
+            return this
+        }
+
+        override fun searchByTitle(title: String): BookQuery {
+            this.empty = false
+            this.title = title
+            this.isbn = null
+            return this
+        }
+
+        override fun searchByISBN13(isbn13: String): BookQuery {
+            this.empty = false
+            this.title = null
+            this.isbn = isbn13
+            return this
+        }
+
+        override fun withOrdering(ordering: BookOrdering): BookQuery {
+            this.ordering = ordering
+            return this
+        }
+
+        override fun getAll(): CompletableFuture<List<Book>> {
+            if (empty) return CompletableFuture.completedFuture(Collections.emptyList())
+            else if (title != null) {
+                return TODO("Search by title")
+            } else { assert(isbn != null)
+                val url = userISBN2URL(isbn!!)
+                if (url != null) {
+                    return url2json(url)
+                        .thenApplyAsync { listOf(parseBook(it)) }
+                        .exceptionally { exception ->
```

Comment:
is `.thenApplyAsync` really what you want? I am 95% sure that the difference between thenApply and thenApplyAsync is that the latter creates a new thread for the block

Reply:
isn't thenApply() synchronous, akin to calling get() ?
I don't understand what other distinction there could be between the two otherwise. 
Also, the doc is not clear and I don't really sure how to test that hypothesis

Reply:
I don't know about the doc. But in the youtube video I watched a week ago, they said that the async version creates a new thread while the non-async uses the same one. Definitely different from get tho because this is how you queue asynchronous operations is Futures/promises

Reply:
I changed it now